src/main/java/org/spongepowered/api/event/network/ClientConnectionEvent.java

```diff
@@ -84,11 +87,39 @@
 
     /**
      * Called after the client authenticates and attempts to login to the
-     * server.
+     * server. This is the phase where plugins can perform a handshake with
+     * the client by sending login related messages and requests.
+     *
+     * <p>During this event, it's possible to use the {@link RawLoginDataChannel}
+     * and the {@link MessageChannel} to send requests to the client. As long as
+     * there's requests going to the client, the connection will stay in the
+     * handshake phase and will not continue to the {@link Login} event.</p>
+     *
+     * <p>For example, a plugin sends a message to the client to request its
+     * client side plugin version. The client responds and the plugin handles
+     * the response. If the plugin decides to send another message, the plugin
+     * handshake phase will stay active. If the plugin doesn't send a message,
+     * it can assumed that the plugin handshake is finished.</p>
+     *
+     * <p>During the lifetime of the handshake phase, a {@link ClientConnection}
+     * can be terminated by calling {@link ClientConnection#disconnect(Text)}.</p>
+     */
+    interface Handshake extends ClientConnectionEvent {
+
+        /**
+         * Gets the {@link ClientConnection}.
+         *
+         * @return The client connection
+         */
+        ClientConnection getConnection();
```

Comment:
If we're a ClientConnectionEvent, should this not be in the root event?

Reply:
I can move it, but it was originally added to the subclasses.